storage/grpc_client.go

```diff
@@ -964,3 +1029,301 @@ func (r *gRPCReader) reopenStream() (*storagepb.ReadObjectResponse, error) {
+				offset, err = w.determineOffset(start)
+				if err == nil {
+					continue
+				}
+			}
+		}
+		if err != nil {
+			return nil, 0, false, err
+		}
+
+		// Update the immediate stream's sent total and the upload offset with
+		// the data sent.
+		sent += len(data)
+		offset += int64(len(data))
+
+		// Not done sending data, do not attempt to commit it yet, loop around
+		// and send more data.
+		if recvd-sent > 0 {
+			continue
+		}
+
+		// Done sending data. Close the stream to "commit" the data sent.
+		resp, finalized, err := w.commit()
+		// Retriable errors mean we should start over and attempt to
+		// resend the entire buffer via a new stream.
+		// If not retriable, falling through will return the error received
+		// from closing the stream.
+		if shouldRetry(err) {
+			sent = 0
+			finishWrite = false
+			offset, err = w.determineOffset(start)
+			if err == nil {
+				continue
+			}
+		}
+		if err != nil {
+			return nil, 0, false, err
+		}
+
+		return resp.GetResource(), offset, finalized, nil
+	}
+}
+
+// determineOffset either returns the offset given to it in the case of a simple
+// upload, or queries the write status in the case a resumable upload is being
+// used.
+//
+// This is an experimental API and not intended for public use.
+func (w *gRPCWriter) determineOffset(offset int64) (int64, error) {
+	// For a Resumable Upload, we must start from however much data
+	// was committed.
+	if w.upid != "" {
+		committed, err := w.queryProgress()
+		if err != nil {
+			return 0, err
+		}
+		offset = committed
+	}
+	return offset, nil
+}
+
+// commit closes the stream to commit the data sent and potentially receive
+// the finalized object if finished uploading. If the last request sent
+// indicated that writing was finished, the Object will be finalized and
+// returned. If not, then the Object will be nil, and the boolean returned will
+// be false.
+//
+// This is an experimental API and not intended for public use.
+func (w *gRPCWriter) commit() (*storagepb.WriteObjectResponse, bool, error) {
+	finalized := true
+	resp, err := w.stream.CloseAndRecv()
+	if err == io.EOF {
+		// Closing a stream for a resumable upload finish_write = false results
+		// in an EOF which can be ignored, as we aren't done uploading yet.
+		finalized = false
+		err = nil
+	}
+	// Drop the stream reference as it has been closed.
+	w.stream = nil
+
+	return resp, finalized, err
+}
+
+// writeObjectSpec constructs a WriteObjectSpec proto using the Writer's
+// ObjectAttrs and applies its Conditions. This is only used for gRPC.
+//
+// This is an experimental API and not intended for public use.
+func (w *gRPCWriter) writeObjectSpec() (*storagepb.WriteObjectSpec, error) {
+	// To avoid modifying the ObjectAttrs embeded in the calling writer, deref
+	// the ObjectAttrs pointer to make a copy, then assign the desired name to
+	// the attribute.
+	attrs := *w.attrs
+	attrs.Name = w.object
```

Comment:
I'm a little confused by this. attrs.Name is already populated by NewWriter. https://github.com/googleapis/google-cloud-go/blob/main/storage/storage.go#L1144 . So why is this necessary?

Reply:
Ah ok, good. I was also confused because the in reviewing the existing HTTP code, it instead [references the ObjectHandle.object directly](https://github.com/googleapis/google-cloud-go/blob/main/storage/writer.go#L174) instead of the attrs. I can change this and elsewhere to use expect the attrs.Name to be set.

Reply:
Note: I won't change the existing http writer code in the prod code path.

Reply:
Cool, sounds good!